app/metal-controller-manager/internal/ipxe/ipxe_server.go

```diff
@@ -224,33 +184,121 @@ func parseMAC(s string) (net.HardwareAddr, error) {
 	return macAddr, err
 }
 
-// determineEnvionment handles which env CRD we'll respect for a given server.
+// newEnvironment handles which env CRD we'll respect for a given server.
 // specied in the server spec overrides everything, specified in the server class overrides default, default is default :).
-func determineEnvironment(c client.Client, serverObj *metalv1alpha1.Server, envObj *metalv1alpha1.Environment) error {
-	envName := "default"
-
-	if serverObj.Spec.EnvironmentRef != nil {
-		envName = serverObj.Spec.EnvironmentRef.Name
-	} else if serverObj.OwnerReferences != nil {
-		// search for serverclass in owner refs. if found, fetch it and see if it's got an env ref
-		for _, owner := range serverObj.OwnerReferences {
-			if owner.Kind == "ServerClass" {
-				serverClassResource := &metalv1alpha1.ServerClass{}
-
-				if err := c.Get(context.Background(), types.NamespacedName{Namespace: "", Name: owner.Name}, serverClassResource); err != nil {
-					return err
-				}
+func newEnvironment(c client.Client, server *metalv1alpha1.Server) (env *metalv1alpha1.Environment, err error) {
+	// NB: The order of this switch statement is important. It defines the
+	// precedence of which environment to boot. The precedence is:
+	// - boot into the agent for registration (if needed)
+	// - boot into the agent for a reset (if needed)
+	// - boot into the environment defined by the server
+	// - boot into the environment defined by the server class
+	// - boot the default environment
+	switch {
+	case server == nil:
+		log.Printf("Unknown server, sending %q option to agent", constants.AgentRegisterArg)
+		env = newAgentEnvironment(constants.AgentRegisterArg)
+	case server.Status.NeedsWipe:
+		log.Printf("Server %q needs wipe, sending %q option to agent", server.Name, constants.AgentResetArg)
+		env = newAgentEnvironment(constants.AgentResetArg)
+	case server.Spec.EnvironmentRef != nil:
+		env, err = newEnvironmentFromServer(c, server)
+		if err != nil {
+			return nil, err
+		}
+	case server.OwnerReferences != nil:
+		env, err = newEnvironmentFromServerClass(c, server)
+		if err != nil {
+			return nil, err
+		}
+	default:
+		env, err = newDefaultEnvironment(c)
```

Comment:
if we check here for `server.InUse` we can return 404 on attempt to boot server which is not use, this should resolve #53